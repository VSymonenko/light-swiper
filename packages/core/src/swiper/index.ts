import {
  CSSSelector,
  ISwiper,
  Slide,
  SwiperOptions as Options,
} from '../../../types';
import {
  detectSelector,
  getElement,
  validateSelector,
} from '../../../utils/src';
import SwiperOptions from './initOptions';

class Swiper extends SwiperOptions implements ISwiper {
  readonly selector: CSSSelector;

  readonly $el!: HTMLElement | Element | null;

  readonly slideWrapper!: HTMLDivElement;

  constructor(selector: CSSSelector, slides?: Slide[], options?: Options) {
    super();
    validateSelector(selector);
    this.selector = selector;
    const type = detectSelector(selector);
    this.$el = getElement(type, selector);
    this.slides = slides || [];
    this.slideWrapper = this.createSlideWrapper();
    if (options) {
      Object.keys(options).forEach((key) => {
        this[key] = options[key];
      });
    }
    this.init();
  }

  // ANCHOR hooks lifecycle
  private init(): void {
    // @ts-ignore-next-line
    // eslint-disable-next-line no-underscore-dangle
    this.$el.__SWIPER__ = this;
  }

  // ANCHOR methods

  private createSlideWrapper(): HTMLDivElement {
    const wrapper = document.createElement('div');
    wrapper.className = 'swiper-wrapper';
    this.$el?.appendChild(wrapper);
    return wrapper;
  }
}

export default Swiper;
