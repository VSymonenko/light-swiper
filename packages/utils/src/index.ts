export * from './selector';
export * from './validator';
