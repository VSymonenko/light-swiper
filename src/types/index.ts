export type Effect = 'slide' | 'fade';

export interface Pagination {
  clickable: boolean;
  activeClass: string;
}

export interface Navigation {
  icons?: {
    next: string;
    previous: string;
  },
  calibration?: number;
}

export interface Breakpoints {
  [key: number]: {
    slidesPerView?: number;
    spaceBetweenSlides?: number;
  }
}

export interface SwiperOptions {
  lock?: boolean;
  anchor?: string;
  effect?: Effect;
  sensitivity?: number;
  activeClass?: string;
  fullWidth?: boolean;
  fullHeight?: boolean;
  width?: number;
  height?: number;
  loop?: boolean;
  isMobile?: boolean;
  navigation?: Navigation;
  pagination?: Pagination;
  slidesPerView?: number;
  spaceBetweenSlides?: number;
  speed?: number
  breakpoints?: Breakpoints;
}

export type CSSSelector = string;

export type CSSSelectorType = 'class' | 'id';

export interface ISwiper extends SwiperOptions {
}
