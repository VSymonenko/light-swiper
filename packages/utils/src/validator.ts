import { CSSSelector } from '@/types';

const startChar = ['.', '#'];
const startMessage = String
  .raw`Selector should be start with '.' or '#' symbol`;

/**
 * If value not a sting type throw type error
 * @param {string} value
 * @returns true | throw new TypeError
 */
export const isString = (value: string) => {
  if (typeof value === 'string') {
    return true;
  }
  throw new TypeError(`Value '${value}' should be a string type`);
};

/**
 * Trhow error if a value is ont CSSSelector
 * @param {CSSSelector} value
 * @returns true | throw new Error
 */
export const validateSelector = (value: CSSSelector) => {
  isString(value);
  const valid = startChar.some((x) => !value.indexOf(x));
  if (valid) {
    return valid;
  }
  throw new Error(startMessage);
};
