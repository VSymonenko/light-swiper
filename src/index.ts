// import initOptions from '@/initOptions';
import { ISwiper } from '@/types';

class Swiper implements ISwiper {
  lock = false;
  // #options: SwiperOptions = initOptions;

  // super(options: SwiperOptions) {
  //   this.#options = options;
  // }
}

export default Swiper;
