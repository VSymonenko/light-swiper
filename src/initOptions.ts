import { SwiperOptions as Options, Effect } from '@/types';

export default class SwiperOptions implements Options {
  lock = false;

  anchor = '';

  effect: Effect = 'slide';

  sensitivity = 1;

  activeClass = 'active';

  fullWidth = true;

  fullHeight = true;

  width = 300;

  height = 300;

  loop = false;

  isMobile = false;

  navigation = {
    icons: {
      next: '>',
      previous: '<',
    },
    calibration: 15,
  };

  pagination = {
    activeClass: 'active-page',
    clickable: true,
  };

  slidesPerView = 1;

  spaceBetweenSlides = 15;

  speed = 300;
}
