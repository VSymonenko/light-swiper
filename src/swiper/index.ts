import SwiperOptions from '@/swiper/initOptions';
import { ISwiper, SwiperOptions as Options, CSSSelector } from '@/types';
// import * as t from 'io-ts';
// import { isRight, fold } from 'fp-ts/lib/Either';
// import { pipe } from 'fp-ts/lib/pipeable';

// const str = new t.Type<string, string, unknown>(
//   'string',
//   (input: unknown): input is string => typeof input === 'string',
//   // `t.success` and `t.failure` are helpers used to build `Either` instances
//   (input, context) => (typeof input === 'string' ? t.success(input) : t.failure(input, context)),
//   // `A` and `O` are the same, so `encode` is just the identity function
//   t.identity,
// );

// // failure handler
// const onLeft = (errors: t.Errors): string => `${errors.length} error(s) found`;

// // success handler
// const onRight = (s: string) => `No errors: ${s}`;

class Swiper extends SwiperOptions implements ISwiper {
  protected selector!: CSSSelector;

  constructor(selector: CSSSelector, options?: Options) {
    super();
    // pipe(t.string.decode(selector), fold(onLeft, onRight));
    this.selector = selector;
    if (options) {
      Object.keys(options).forEach((key) => {
        this[key] = options[key];
      });
    }
  }
}

export default Swiper;
