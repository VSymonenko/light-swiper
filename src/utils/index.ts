import { CSSSelector, CSSSelectorType } from '@/types';

/**
 * return specifyed selector type
 * @param {CSSSelector} selector
 * @returns {CSSSelectorType} 'id' | 'class'
 */
export const detectSelector = (
  selector: CSSSelector,
): CSSSelectorType => (!selector.indexOf('#') ? 'id' : 'class');

/**
 * return DOM node by selector
 * @param {CSSSelectorType} type
 * @param {CSSSelector} selector
 * @returns {(HTMLElement | Element | null)} HTMLElement | Element | null
 */
export const getElement = (
  type: CSSSelectorType, selector: CSSSelector,
): HTMLElement | Element | null => (
  type === 'id' ? document.getElementById(selector)
    : document.getElementsByClassName(selector).item(0)
);
