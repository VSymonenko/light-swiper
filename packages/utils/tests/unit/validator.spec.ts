import { isString, validateSelector } from '../../src';

describe('should be check string type', () => {
  test('should be pass string', () => {
    const str = 'test string';
    expect(isString(str)).toBeTruthy();
    expect(isString(str)).toMatchSnapshot();
  });

  test('should be thtrow error when pass wrong type argument', () => {
    const types = [
      null,
      undefined,
      0,
      BigInt(Number.MAX_SAFE_INTEGER),
      true,
      Symbol('test'),
      {},
      ['array'],
    ];
    types.forEach((type) => {
      expect(() => {
        // @ts-ignore
        expect(isString(type));
      }).toThrowError();

      expect(() => {
        // @ts-ignore
        expect(isString(type));
      }).toThrowErrorMatchingSnapshot();
    });
  });
});

describe('should be validate selector', () => {
  test('should be valid id type selector', () => {
    const id = '#swiper';
    expect(validateSelector(id)).toBeTruthy();
    expect(validateSelector(id)).toMatchSnapshot();
  });

  test('shouled be valid class type selector', () => {
    const className = '.swiper';
    expect(validateSelector(className)).toBeTruthy();
  });
  test('should be throw error', () => {
    const wrongSelector = 'wrong-selector';
    expect(() => validateSelector(wrongSelector)).toThrowError();
    expect(() => {
      validateSelector(wrongSelector);
    }).toThrowErrorMatchingSnapshot();
  });
});
