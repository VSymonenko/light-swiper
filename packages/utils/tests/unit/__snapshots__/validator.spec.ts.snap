// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be check string type should be pass string 1`] = `true`;

exports[`should be check string type should be thtrow error when pass wrong type argument 1`] = `"Value 'null' should be a string type"`;

exports[`should be check string type should be thtrow error when pass wrong type argument 2`] = `"Value 'undefined' should be a string type"`;

exports[`should be check string type should be thtrow error when pass wrong type argument 3`] = `"Value '0' should be a string type"`;

exports[`should be check string type should be thtrow error when pass wrong type argument 4`] = `"Value '9007199254740991' should be a string type"`;

exports[`should be check string type should be thtrow error when pass wrong type argument 5`] = `"Value 'true' should be a string type"`;

exports[`should be check string type should be thtrow error when pass wrong type argument 6`] = `"Cannot convert a Symbol value to a string"`;

exports[`should be check string type should be thtrow error when pass wrong type argument 7`] = `"Value '[object Object]' should be a string type"`;

exports[`should be check string type should be thtrow error when pass wrong type argument 8`] = `"Value 'array' should be a string type"`;

exports[`should be validate selector should be throw error 1`] = `"Selector should be start with '.' or '#' symbol"`;

exports[`should be validate selector should be valid id type selector 1`] = `true`;
